crates/holochain_sqlite/src/db/access.rs

```diff
@@ -337,13 +352,22 @@ impl<Kind: DbKindT + Send + Sync + 'static> DbWrite<Kind> {
             log_elapsed!([10, 100, 1000], start, "write_async:before-closure");
             let r = conn.execute_in_exclusive_rw_txn(f);
             log_elapsed!([10, 100, 1000], start, "write_async:after-closure");
-            r
+            r.map(|r| (r, permit))
         }).in_current_span()).in_current_span().await.map_err(|e| {
             tracing::error!("Failed to claim a thread to run the database write transaction. It's likely that the program is out of threads.");
             DatabaseError::Timeout(e)
         })?.map_err(DatabaseError::from)?
     }
 
+    /// Acquire the single write permit for the database.
+    ///
+    /// This will prevent any other writes from proceeding until the semaphore is released.
+    /// It can be used to ensure that a read, followed by other operations, followed by another
+    /// write, will not be interleaved with some other write.
+    pub async fn acquire_write_permit(&self) -> DatabaseResult<OwnedSemaphorePermit> {
+        acquire_semaphore_permit(self.0.write_semaphore.clone()).await
+    }
```

Comment:
I'm not sure how we ended up with semaphores on top of the native database locking. It feels like a duplication of functionality. However, I don't think we should change it at this point, so this looks like a good solution with what we have to work with.

Reply:
I know I've seen documentation somewhere that recommends limiting concurrent access with something like this but I can't find it so I can't double check what the advice actually was and whether it's relevant. It is useful for things like this, and the cascade though.

I'm not convinced by the implementation in the cascade but there are TODOs there for that, as in so many places :)